libraries/botbuilder-repo-utils/src/execCmd.ts

```diff
@@ -2,31 +2,98 @@
 // Licensed under the MIT License.
 
 import * as R from 'remeda';
+import assert from 'assert';
+import async from 'async';
 import minimist from 'minimist';
 import path from 'path';
-import { DependencyResolver, collectWorkspacePackages } from './workspace';
+import pmap from 'p-map';
 import { Package } from './package';
 import { Result, failure, run, success } from './run';
-import { execute, stream } from './process';
+import { Workspace, collectWorkspacePackages } from './workspace';
+import { execute } from './process';
 import { gitRoot } from './git';
 import { readJsonFile } from './file';
 
+// Describes a function signature for an executor that accepts a list of workspaces,
+// a task, and a concurrency parameter and executes the task against all workspaces.
+type Executor = (
+    workspaces: Workspace[],
+    task: (workspace: Workspace) => Promise<unknown>,
+    concurrency?: number
+) => Promise<unknown>;
+
+// Executes `task` against `workspaces` ensuring that `task` is executed against dependencies
+// before consumers (if possible, cycles are broken).
+const coordinatedExecutor: Executor = (workspaces, task, concurrency) => {
+    const names = new Set(workspaces.map((workspace) => workspace.pkg.name));
+
+    // Collect packages with all workspace deps
+    const workspacesByPkgName: Record<string, { workspace: Workspace; dependencies: string[] }> = workspaces.reduce(
+        (acc, workspace) => ({
+            ...acc,
+            [workspace.pkg.name]: {
+                workspace,
+                dependencies: R.uniq([
+                    ...Object.keys(workspace.pkg.dependencies ?? {}).filter((name) => names.has(name)),
+                    ...Object.keys(workspace.pkg.devDependencies ?? {}).filter((name) => names.has(name)),
+                ]),
+            },
+        }),
+        {}
+    );
+
+    // Construct a DAG of tasks, with cycles removed, for async execution
+    const tasks = Object.entries(workspacesByPkgName).reduce((acc, [, { dependencies, workspace }]) => {
+        return {
+            ...acc,
+            [workspace.pkg.name]: [
+                ...R.reject(dependencies, (name) =>
+                    // eslint-disable-next-line security/detect-object-injection
+                    workspacesByPkgName[name].dependencies.includes(workspace.pkg.name)
+                ),
+                (...args: unknown[]) => {
+                    const callback = args.pop();
+                    assert(typeof callback === 'function');
+
+                    task(workspace)
+                        .then(() => callback())
+                        .catch((err) => callback(err));
+                },
+            ],
+        };
+    }, {});
+
+    return new Promise((resolve, reject) => async.auto(tasks, concurrency, (err) => (err ? reject(err) : resolve())));
```

Comment:
`async.auto` is designed for exactly this problem.

The main reason I removed my old implementation is that it suffered from "head-of-line blocking". In particular, tasks were executed concurrently in groups. As a consequence, the slowest task in a group determines how long it takes to execute the group of tasks.

`async.auto` ensures that, if possible, `N` tasks are always being executed, where `N = concurrency`.